documentation/decisions/adr/000n-backend-tooling.md

```diff
@@ -0,0 +1,92 @@
+
+## Decision Drivers <!-- RECOMMENDED -->
+
+- **Ease of use and configurability:** Code quality tools should be configurable to meet the specific needs of our project and enforce standards uniformly. Running the tool should be simple, and feedback should be easily implementable and understandable.
+- **Speed:** Code quality tools should be able to run in an timely manner, ideally in a pre-commit hook.
+- **Documentation and resources:** Code quality tools should have robust and helpful documentation, specifically around usage and configuration.
+- **Lean:** Selection of code quality tools should cover all use-cases with little overlap for ease of use and to minimize mental surface area required to understand the system. As in, each tool should have a specific, meaningful purpose.
+- **Enforces security:** Tooling should assist us in identifying security concerns and vulnerabilities.
+- **Well-maintained:** We have a preference towards libraries that are widely adopted and have active maintainers.
+
+## Options Considered
+
+1. Use a curated collection of Python libraries (described below)
+2. Use a different or modified set of Python libraries
+
+### Dependency Management
+
+**[Poetry](https://python-poetry.org/docs/):** Python packaging and dependency management.  
+### Code Linting
+
+**[Flake8](https://flake8.pycqa.org/en/latest/#):** Format and enforce style in our Python code.  
+
+*Additional Flake Extensions:*
+- [bugbear](https://pypi.org/project/flake8-bugbear/) for finding likely bugs and design problems.
+- [alfred](https://pypi.org/project/flake8-alfred/) for finding unsafe/obsolete symbols.
+
+### Auto-formatting
+
+**[isort](https://pycqa.github.io/isort/):** Sort Python imports.  
+
+**[Black](https://github.com/psf/black):** Format Python code.  
+
+### Type Checking
+
+**[Mypy](https://mypy-lang.org/):** Validate and enforce static type checking in Python.  
+### Security
+
+**[Bandit](https://bandit.readthedocs.io/en/latest/):** Security checking tool used to identify common concerns in Python code.  
+
+**[Safety](https://docs.pyup.io/docs/getting-started-with-safety-cli):** Safety first! Safety scans dependencies for vulnerabilities and security concerns.[^*]  
+
+### License Checking[^*]
+
+**[pip-licenses](https://github.com/raimon49/pip-licenses):** CLI tool for checking the software license of installed Python packages with `pip`.  
+
+### Interface
+
+**[Make](https://www.gnu.org/software/make/manual/make.html):** Run scripts, linters and formatters.  
+
+## Decision Outcome <!-- REQUIRED -->
+
+TBD.
+
+## Other Options
+
+Adopting [Tox](https://tox.wiki/en/latest/) as a testing / linting manager with some of the libraries.
+
+**Dependency Management:**  
+[Pipenv](https://pipenv.pypa.io/en/latest/)
+
+**Code Linting:**  
+[Ruff](https://github.com/astral-sh/ruff) was also considered as an option for its speed and growing community adoption.  
```

Comment:
I'll defer to you all for the final choice, but personally I would lean toward Ruff over Flake8:

- It's faster
- It has more GitHub stars (16k vs <3k for flake8)
- It [re-implements](https://beta.ruff.rs/docs/faq/#how-does-ruff-compare-to-flake8) `isort` and `flake8-bugbear` natively so we wouldn't need separate dependencies for those

This is where even a 1-sentence description on why a certain option was chosen over another would be helpful. Though as we talked about, linting choices are fairly low stakes, so don't let that be a blocker!

Reply:
I wanted to add that the template has configuration files set up already as a starting point. Would want to verify that switching to Ruff isn't a lot of increased overhead to set up and that it doesn't conflict with other packages listed.

Reply:
The footnote references everything implemented in the template repository as the previously curated collection, including Flake8. Since there are a lot of tools to decide on, I was indexing on brevity and on using the template as an MVP to get spun up quickly. Seems like it's worthwhile to switch to Ruff for our project! It's fancier in every way it seems.

The template has configuration files set up already as a starting point. Would want to verify that switching to Ruff isn't a lot of increased overhead to set up and that it doesn't conflict with other packages listed (although we'd be able to get rid of some! 🍾 ). Might open a PR against our template to check that out.

Reply:
I approved this as is, since we have some working examples with non-ruff tools. @daphnegold Will leave it your discretion whether to wait a ruff PR or pull in as-is for now.

Reply:
Yeah don't let my notes be a blocker for getting something merged in and adopting these tools. The ones you suggested are smart defaults and we can always iterate!

Reply:
I modified the ADR to adopt Ruff because it seems super awesome. Did so with the caveat that it will require some effort to move away from the set of packages implemented by the Flask template repository. In my brief research, I don't think it will be that difficult. Someone built a tool to migrate Flake8 => Ruff. It is compatible out of the box with Black. 

Unsure right now whether to do the work in the Flask template repository or adopt it only for this project with Flake8 initially and then a follow up to migrate to Ruff. 🤷‍♀️